### ISSUE TITLE
[RLlib] Policy migration TODOs

### ISSUE THREAD

**Issue by gjoliver** [ray-project/ray#24806]
### Description

We are migrating RLlib's policy implementation to be sub-class based. The remaining tasks in the order priority are:

Finish migrating the rest of the agents. [P0]

Removing references of worker and worker_index from Policy classes. [P1]
Policy right now uses worker_index to decide whether a model should be placed on CPU or GPU.
This type of logic needs to happen in application layer and gets passed in.

Be able to by pass loss and optimizer initialization for inference. [P1]
This includes by-passing loss and view_requirements initialization, MultiGPUTower creation etc.
Basically, things that are only relevant for training should only happen in training stack.
View requirements for inference should just be obs plus all the state_in/out heads.

Move training related states out of Policy. [P2]
Even nicer would be to keep Policy as an extremely thin layer around model.
All training-only bits would live somewhere in Trainer and Rollout worker.
This would significantly improve the shelf-life of our trained Policies.

### Use case

RLlib policy serving.

**Comment by gjoliver** [comment]
@smorad and @kouroshHakha also have a lot of good comments on the PR https://github.com/ray-project/ray/pull/24742
centering around better API names, consolidating API calls, and some potential bugs.
we should converge on the decisions and tackle them as a team too.

**Comment by smorad** [comment]
Quick summary of the biggest points:

1. 
> For a user trying to implement their own model forward pass, it's quite confusing we have compute_actions_from_input_dict, _compute_action_helper, compute_actions, and action_sampler_fn.
>
> Could any of these be merged together or named differently to help the users understand which one they want to override? It's not clear at first glance.

2.
> Does it make sense to deprecate learn_on_batch and use learn_on_loaded_batch for the single-GPU/CPU case as well?

3.
> Using `torch.jit.trace` to save models will trace a model WITHOUT INSPECTING BRANCHES. We should definitely not use this to export torch models, as even TorchModelV2 has if/else statements in the forward pass. Use `torch.save/torch.load` to pickle/unpickle is probably much safer. If the user wants to `JIT`, let them do it themselves.

4.
> Taking the mean loss over towers is probably NOT what we want. We likely want the sum instead, see below for example:
>
> Let's say we have a rollout of three timesteps that produces losses 0.1, 0.2, 0.3. If we were to train the entire rollout using a single tower, the loss would be loss = 0.1 + 0.2 + 0.3 = 0.6. However, now let's say we compute the same exact losses using three towers -- we split up the rollout and put one timestep on each tower. If we take the mean over the three towers, then: loss = (0.1 + 0.2 + 0.3) / 3 = 0.2.
>
> So if we take the mean, then increasing the number of towers will actually "slow down" the learning, by reducing the lr to lr / num_towers . So if you change the number of towers, you would need to multiply the lr by num_towers to retain the same timestep efficiency and training results.

**Comment by gjoliver** [comment]
Post migration clean up:

* consolidate all the action computation APIs to the one and only compute_actions()
* find a consistent naming scheme for all the APIs users can override, for example we have loss() and optimizer(), and we also have all the functions that ends with _fn(). we are keeping the names for now so users know how to migrate their code.
* TF and Torch policies have different APIs for model and action_dist making today. make_model() and make_model_and_action_dist(). consolidate them.

**Comment by sven1977** [comment]
Left some comments and answers on the PR (it's already merged to not delay things further as this was a huge, risky PR that I wanted to merge asap): 

On @smorad 's points above:

1. and 2.: All learning/updating logic will be removed from the Policy classes once we integrate Ray Train into RLlib (a project, we are currently working on and soon will have a prototype for). RLlib should only use a simple update/train API on each policies/models and the update logic should move into the realm of Ray Train (distributed updates). This would also solve our multi-GPU problems (all handled by Ray Train then) and make our policies slimmer.

4. Yes, we SHOULD really change API names for Trainer and Policy classes. But not in this very project. We should have a thorough internal discussion on which names we should change and keep the old names supported (with warning) for some time. This is something we only want to do once to not further confuse users. See this PR here for a first grounds of discussion:


https://github.com/ray-project/ray/pull/25293

**Comment by gjoliver** [comment]
please don't close
